linera-views/src/backends/lru_caching.rs

```diff
@@ -55,60 +55,222 @@ mod metrics {
         match self {
-            CacheEntry::Value(vec) => vec.len(),
+            ValueCacheEntry::Value(vec) => vec.len(),
             _ => 0,
         }
     }
 }
 
+struct FindKeysEntry(BTreeSet<Vec<u8>>);
+
+impl FindKeysEntry {
+    fn size(&self) -> usize {
+        self.0.iter().map(|key| key.len()).sum()
+    }
+
+    fn get_keys_by_prefix(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|key| key[delta..].to_vec())
+            .collect()
+    }
+
+    fn contains_key(&self, key: &[u8]) -> bool {
+        self.0.contains(key)
+    }
+
+    fn update_entry(&mut self, key: &[u8], new_value: Option<Vec<u8>>) {
+        match new_value {
+            None => {
+                self.0.remove(key);
+            }
+            Some(_) => {
+                self.0.insert(key.to_vec());
+            }
+        }
+    }
+
+    fn delete_prefix(&mut self, key_prefix: &[u8]) {
+        let keys = self
+            .0
+            .range(get_interval(key_prefix.to_vec()))
+            .cloned()
+            .collect::<Vec<_>>();
+        for key in keys {
+            self.0.remove(&key);
+        }
+    }
+}
+
+struct FindKeyValuesEntry(BTreeMap<Vec<u8>, Vec<u8>>);
+
+impl FindKeyValuesEntry {
+    fn size(&self) -> usize {
+        self.0
+            .iter()
+            .map(|(key, value)| key.len() + value.len())
+            .sum()
+    }
+
+    fn get_keys_by_prefix(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|(key, _)| key[delta..].to_vec())
+            .collect()
+    }
+
+    fn get_find_key_values(&self, key_prefix: &[u8]) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let key_prefix = key_prefix.to_vec();
+        let delta = key_prefix.len();
+        self.0
+            .range(get_interval(key_prefix))
+            .map(|(key, value)| (key[delta..].to_vec(), value.to_vec()))
+            .collect()
+    }
+
+    fn contains_key(&self, key: &[u8]) -> bool {
+        self.0.contains_key(key)
+    }
+
+    fn get_read_value(&self, key: &[u8]) -> Option<Vec<u8>> {
+        self.0.get(key).cloned()
+    }
+
+    fn update_entry(&mut self, key: &[u8], new_value: Option<Vec<u8>>) {
+        match new_value {
+            None => {
+                self.0.remove(key);
+            }
+            Some(new_value) => {
+                self.0.insert(key.to_vec(), new_value);
+            }
+        }
+    }
+
+    fn delete_prefix(&mut self, key_prefix: &[u8]) {
+        let keys = self
+            .0
+            .range(get_interval(key_prefix.to_vec()))
+            .map(|(key, _)| key.clone())
+            .collect::<Vec<_>>();
+        for key in keys {
+            self.0.remove(&key);
+        }
```

Comment:
if you iterate over the result of the `self.0.range.map` you won't have to allocate the vector